File: Data Structures & Algorithms/count-number-of-islands/submission-3.kt
class Solution {
    fun numIslands(grid: Array<CharArray>): Int {
        val m = grid.size
        val n = grid[0].size

        fun dfs(i: Int,j: Int) {
            if((i !in 0 until m ) || (j !in 0 until n) || grid[i][j] != '1')
                return
            grid[i][j] = '0'
            dfs(i-1,j)
            dfs(i+1,j)
            dfs(i,j+1)
            dfs(i,j-1)
        }
        var res = 0
        for(i in 0 until m)
            for (j in 0 until n)
                if (grid[i][j] == '1'){
                    res += 1
                    dfs(i,j)
                }
        return res
    }
}

File: Data Structures & Algorithms/surrounded-regions/submission-3.py
class Solution:
    def solve(self, grid: List[List[str]]) -> None:
        dir = [(1, 0), (0, 1), (-1, 0), (0, -1)]
        m, n = len(grid), len(grid[0])
        def dfs(i, j):
            grid[i][j] = "S"
            for d in dir:
                ni, nj = i + d[0], j + d[1]
                if 0 <= ni < m and 0 <= nj < n and grid[ni][nj] == "O":
                    dfs(ni, nj)
        

        for i in range(m):
            if grid[i][0] == "O":
                dfs(i, 0)
            if grid[i][n-1] == "O":
                dfs(i, n-1)
        for j in range(n):
            if grid[0][j] == "O":
                dfs(0, j)
            if grid[m-1][j] == "O":
                dfs(m-1, j)
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "O":
                    grid[i][j] = "X"
                if grid[i][j] == "S":
                    grid[i][j] = "O"